Sources/Testing/Traits/AttachmentSavingTrait.swift

```diff
@@ -0,0 +1,337 @@
+  public struct Condition: Sendable {
+    /// The testing library saves attachments if the test passes.
+    public static var testPasses: Self {
+      Self { !$0.hasFailed }
+    }
+
+    /// The testing library saves attachments if the test fails.
+    public static var testFails: Self {
+      Self { $0.hasFailed }
+    }
+
+    /// The testing library saves attachments if the test records a matching
+    /// issue.
+    ///
+    /// - Parameters:
+    ///   - issueMatcher: A function to invoke when an issue occurs that is used
+    ///     to determine if the testing library should save attachments for the
+    ///     current test.
+    ///
+    /// - Returns: An instance of ``AttachmentSavingTrait/Condition`` that
+    ///   evaluates `issueMatcher`.
+    public static func testRecordsIssue(
+      matching issueMatcher: @escaping @Sendable (_ issue: Issue) async throws -> Bool
+    ) -> Self {
+      Self(inspectsIssues: true) { context in
+        for issue in context.issues {
+          if try await issueMatcher(issue) {
+            return true
+          }
+        }
+        return false
+      }
+    }
+
+    /// Whether or not this condition needs to inspect individual issues (which
+    /// implies a slower path.)
+    fileprivate var inspectsIssues = false
+
+    /// The condition function.
+    ///
+    /// - Parameters:
+    ///   - condition: The function to call. The result of this function
+    ///     determines if the condition is satisfied or not.
+    fileprivate var body: @Sendable (borrowing Context) async throws -> Bool
+  }
+
+  /// This instance's condition.
+  var condition: Condition
+
+  /// The source location where this trait is specified.
+  var sourceLocation: SourceLocation
+
+  public var isRecursive: Bool {
+    true
+  }
+}
+
+// MARK: - TestScoping
+
+extension AttachmentSavingTrait: TestScoping {
+  /// A type representing the per-test-case context for this trait.
+  ///
+  /// An instance of this type is created for each scope this trait provides.
+  /// When the scope ends, the context is then passed to the trait's condition
+  /// function for evaluation.
+  fileprivate struct Context: Sendable {
+    /// The set of events that were deferred for later conditional handling.
+    var deferredEvents = [Event]()
+
+    /// Whether or not the current test case has recorded a failing issue.
+    var hasFailed = false
+
+    /// All issues recorded within the scope of the current test case.
+    var issues = [Issue]()
+  }
+
+  public func scopeProvider(for test: Test, testCase: Test.Case?) -> Self? {
+    // This function should apply directly to test cases only. It doesn't make
+    // sense to apply it to suites or test functions since they don't run their
+    // own code.
+    //
+    // NOTE: this trait can't reliably affect attachments recorded when other
+    // traits are evaluated (we may need a new scope in the TestScoping protocol
+    // for that.)
+    testCase != nil ? self : nil
```

Comment:
I think in this case, since `isRecursive` is `true`, you can just omit the `scopeProvider()` implementation entirely and it will have the equivalent behavior you have here.

Reply:
I'd prefer it be explicit so I have a place to put the comment, but we can always delete this code later if we find it's redundant.